src/aws/v4.rs

```diff
@@ -213,6 +214,28 @@ impl Signer {
     }
 }
 
+#[async_trait::async_trait]
+impl<T: Send> Sign<T> for Signer {
+    type Credential = Credential;
+
+    async fn sign(
+        &self,
+        req: &mut http::request::Request<T>,
```

Comment:
i found that it's not that easy to get the mut ref of the `Parts` inside `http::request::Request<T>`, to mut the req, it seems that it's still easier to pass `&mut http::request::Request<T>` here if we'd like to pass a concerete `Request` struct here.

Reply:
> i found that it's not that easy to get the mut ref of the `Parts` inside `http::request::Request<T>`, to mut the req

Would you like to elaborate further? It should be as straightforward as `into_parts` and `from_parts`. I want to avoid introduce an generice body in sign.

Reply:
let me have a try again.

Reply:
we have to have the ownership of Request in the caller side, like this:

```
let (mut parts, body) = req.into_parts();
signer.sign(mut parts);
let req = Request::from_parts(parts, body);
```

if all you have is a `&mut Request`, you can not easily get the `parts` from it in the caller site.

Reply:
I proposed an RFC to `http`: https://github.com/hyperium/http/issues/712

--


> we have to have the ownership of Request in the caller side, like this:

In read use cases, caller will always have the ownership of `Request`. I still think it's better for us to just take `Parts` in our API. So all imples, users don't need to play with `Body` generic type.

Reply:
this RFC makes sense.

i've made the changes, PTAL

Reply:
if hyperium/http is not planned to exposes `parts_mut()`, i'm afraid that it might be better api to keep accepting a generic parameter in the `Sign` trait.

in the current code base, there're a lot of call sites that only have got `&mut Request` instead of an owned `Request`. it's hard to get the mutable reference of `Parts`.

Reply:
I strongly disagree with accepting a generic parameter. I prefer refactoring the entire codebase.